Tools/TestWebKitAPI/Tests/mac/LoadWebArchive.mm

```diff
@@ -644,4 +646,46 @@ HTTPServer server({
     doneEvaluatingJavaScript = false;
 }
 
+TEST(LoadWebArchive, PreferCachedImageSourceOverBrokenImage)
+{
+    [TestProtocol registerWithScheme:@"http"];
+
+    RetainPtr archiveData = [] {
+        RetainPtr sourceView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 800, 600)]);
+        [sourceView _setOverrideDeviceScaleFactor:1];
+        [sourceView synchronouslyLoadHTMLString:@"<!DOCTYPE html>"
+            "<html>"
+            "  <meta name='viewport' content='width=device-width, initial-scale=1'>"
+            "  <body>"
+            "    <picture>"
+            "      <img src='http://bundle-file/400x400-green.png'"
+            "        srcset='http://bundle-file/400x400-green.png 1x,http://bundle-file/large-red-square.png 2x' />"
+            "    </picture>"
+            "  </body>"
+            "</html>"];
+        RetainPtr result = [sourceView contentsAsWebArchive];
+        [sourceView _killWebContentProcessAndResetState];
+        return result;
+    }();
+
+    [TestProtocol unregister];
+
+    RetainPtr configuration = adoptNS([WKWebViewConfiguration new]);
+    [configuration setWebsiteDataStore:WKWebsiteDataStore.nonPersistentDataStore];
+    [configuration _setAllowedNetworkHosts:NSSet.set];
+
+    RetainPtr delegate = adoptNS([TestNavigationDelegate new]);
+    RetainPtr webView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 800, 600) configuration:configuration.get()]);
+    [webView setNavigationDelegate:delegate.get()];
+    [webView _setOverrideDeviceScaleFactor:2];
+    [webView loadData:archiveData.get() MIMEType:@"application/x-webarchive" characterEncodingName:@"" baseURL:[NSURL URLWithString:@""]];
+    [delegate waitForDidFinishNavigation];
+
+    NSArray<NSNumber *> *dimensions = [webView objectByEvaluatingJavaScript:@"const image = document.images[0]; [image.naturalWidth, image.naturalHeight];"];
+    auto naturalWidth = dimensions.firstObject.doubleValue;
+    auto naturalHeight = dimensions.lastObject.doubleValue;
+    EXPECT_GT(naturalWidth, 0);
+    EXPECT_GT(naturalHeight, 0);
```

Comment:
I think these conditions will be always true even for broken images. I think they will be { 20, 20 } if the default broken image is drawn. Should not these be { 400, 400 }?

Reply:
I verified that without this fix, the natural width and height are indeed 0 (so the test fails).

However, I can adjust the test to make this check clearer and use `EXPECT_EQ` instead.

Reply:
My memory is broken image is used for image load, and with the filtering the image load will not even start

Reply:
FWIW, the load failure that results in the broken image comes from this codepath:

```
    if (auto* page = frame ? frame->page() : nullptr) {
        auto mainFrameMainResource = frame->isMainFrame()
            && resourceLoader.frameLoader()
            && resourceLoader.frameLoader()->notifier().isInitialRequestIdentifier(identifier)
            ? MainFrameMainResource::Yes : MainFrameMainResource::No;
        if (!page->allowsLoadFromURL(request.url(), mainFrameMainResource)) {
            RunLoop::main().dispatch([resourceLoader = Ref { resourceLoader }, error = blockedError(request)] {
                resourceLoader->didFail(error);
            });
            return;
        }
    }
```

It's true that the image load never actually starts (in the network process, at least) — but this does seem to result in a broken image being shown still.